src/libraries/Native/Unix/System.Native/pal_networking.c

```diff
@@ -440,6 +429,135 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
     return ret;
 }
 
+#if HAVE_GETADDRINFO_A
+struct GetAddrInfoAsyncState
+{
+    struct gaicb gai_request;
+    struct gaicb* gai_requests;
+    struct sigevent sigevent;
+
+    HostEntry* entry;
+    GetHostEntryForNameCallback callback;
+    char address[];
+};
+
+static void GetHostEntryForNameAsyncComplete(sigval_t context)
+{
+    struct GetAddrInfoAsyncState* state = (struct GetAddrInfoAsyncState*)context.sival_ptr;
+
+    atomic_thread_fence(memory_order_acquire);
+
+    GetHostEntryForNameCallback callback = state->callback;
+
+    int ret = ConvertGetAddrInfoAndGetNameInfoErrorsToPal(gai_error(&state->gai_request));
+
+    if (ret == 0)
+    {
+        const uint8_t* address = (const uint8_t*)state->address;
+        struct addrinfo* info = state->gai_request.ar_result;
+
+        ret = GetHostEntries(address, info, state->entry);
+    }
+
+    if (callback != NULL)
```

Comment:
is there valid case when the callback be null?

Reply:
No.

In C# I would write `Debug.Assert(callback is not null)`. Can we do something similar here or just drop the `if`?

Reply:
Will use
```c
    assert(callback != NULL);
```
instead.